File: backend/app/schemas/agent.py
"""Pydantic schemas for the Agent API."""

from __future__ import annotations

from typing import List, Optional

from pydantic import BaseModel, Field


class CreateSessionRequest(BaseModel):
    document: str = Field(default="", description="Initial document content")
    title: str = Field(default="Untitled", description="Initial document title")


class CreateSessionResponse(BaseModel):
    session_id: str
    version: int
    title: str


class ContextItem(BaseModel):
    """A named document snippet attached to a message, referenced via ``@<ref>``.

    The frontend lets users attach multiple snippets from the rendered document
    (each gets an auto-assigned ref like ``ctx-1``) and reference them inline in
    the message with ``@ctx-1``. The backend expands those references into
    fenced Markdown blocks before handing the message to the agent.
    """

    ref: str = Field(
        ...,
        pattern=r"^[A-Za-z0-9][A-Za-z0-9_-]*$",
        max_length=64,
        description="Short reference token used in the message as @<ref>",
    )
    label: str = Field(default="", max_length=200, description="Human-readable snippet label")
    content: str = Field(..., max_length=200_000, description="Raw Markdown snippet content")


class SendMessageRequest(BaseModel):
    message: str = Field(..., min_length=1, description="User message to the agent")
    provider: str = Field(default="deepseek", description="AI provider")
    model: str = Field(..., description="Model name")
    selection: Optional[str] = Field(
        default=None, description="Selected text context (from @selection)"
    )
    cursor_position: int = Field(default=0)
    contexts: Optional[List[ContextItem]] = Field(
        default=None,
        description="Named context snippets referenced in the message via @<ref>",
    )


class ClientSyncRequest(BaseModel):
    base_version: int = Field(..., description="Version the client's edit is based on")
    content: str = Field(..., description="Full new document content from the client")


class ClientSyncResponse(BaseModel):
    status: str = Field(..., description="ok | conflict")
    version: int
    content: str
    title: str


class StopResponse(BaseModel):
    stopped: bool

File: frontend/src/services/types/agent.ts
// Agent SSE event types (mirror backend /api/v1/agent)
// Backend source: backend/app/services/agent/translator.py + service.py

export type AgentEventType =
  | 'thought'
  | 'tool_call'
  | 'tool_result'
  | 'document_patch'
  | 'final'
  | 'stopped'
  | 'error'
  | 'done';

export interface ThoughtEvent {
  type: 'thought';
  content: string;
}

export interface ToolCallEvent {
  type: 'tool_call';
  name: string;
  args: Record<string, unknown>;
}

export interface ToolResultEvent {
  type: 'tool_result';
  name: string;
  ok: boolean;
  summary: string;
}

export interface DocumentPatchEvent {
  type: 'document_patch';
  version: number;
  summary: string;
}

export interface FinalEvent {
  type: 'final';
  content: string;
}

export interface StoppedEvent {
  type: 'stopped';
  content: string;
}

export interface ErrorEvent {
  type: 'error';
  error: string;
}

export interface DoneEvent {
  type: 'done';
  content: string;
}

export type AgentEvent =
  | ThoughtEvent
  | ToolCallEvent
  | ToolResultEvent
  | DocumentPatchEvent
  | FinalEvent
  | StoppedEvent
  | ErrorEvent
  | DoneEvent;

// Request types
export interface CreateSessionRequest {
  document: string;
  title: string;
}

export interface CreateSessionResponse {
  session_id: string;
  version: number;
  title: string;
}

export interface SendMessageRequest {
  message: string;
  provider: string;
  model: string;
  selection?: string;
  cursor_position?: number;
}

export interface ClientSyncRequest {
  base_version: number;
  content: string;
}

export interface ClientSyncResponse {
  status: 'ok' | 'conflict';
  version: number;
  content: string;
  title: string;
}
